### PR TITLE
feat(frontends): Implement rfft in PaddlePaddle frontend and fix fft for Tensorflow backend

### LINKED ISSUE #19383: rfft

### PR DESCRIPTION
**Pull request by AwkNinja**
Close #19383

### CODE REVIEW

**Comment by AwkNinja** [comment]
Still working on fixing the failing tests for tensorflow and torch.

**Comment by AwkNinja** [comment]
@hmahmood24 Can you please have a look at the implementation here. I can't get the tensorflow and torch tests to pass.

**Comment by AwkNinja** [comment]
The torch tests seem to be passing now but still having issues with tensorflow. The main error is that the elements list is empty so it throws the error cannot draw from an empty list. At first I thought it was because of the norm list that I am generating but it doesn't seem like its related to it.

File: ivy_tests/test_ivy/test_frontends/test_paddle/test_fft.py [hunk 1]
```diff
@@ -177,3 +177,33 @@ def test_paddle_ifftshift(dtype_x_axis, frontend, test_flags, fn_tree, on_device
...
+
+# rfft
+@handle_frontend_test(
+    fn_tree="paddle.fft.rfft",
+    dtype_input_axis=helpers.dtype_values_axis(
+        available_dtypes=helpers.get_dtypes("float"),
+        shape=(2,),
+        min_axis=-1,
+        force_int_axis=True,
+    ),
```
Comment (hmahmood24) [inline ivy_tests/test_ivy/test_frontends/test_paddle/test_fft.py:190]: Why can't we use something like this?

```python
dtype_input_axis=helpers.dtype_values_axis(
        available_dtypes=helpers.get_dtypes("valid"),
        min_num_dims=1,
        min_dim_size=2,
        shape=helpers.get_shape(...),
        force_int_axis=True,
    )
```

File: ivy_tests/test_ivy/test_frontends/test_paddle/test_fft.py [hunk 2]
```diff
@@ -177,3 +177,33 @@ def test_paddle_ifftshift(dtype_x_axis, frontend, test_flags, fn_tree, on_device
...
+    norm=st.sampled_from(["backward", "ortho", "forward"]),
+    n=st.integers(min_value=2, max_value=10),
```
Comment (hmahmood24) [inline ivy_tests/test_ivy/test_frontends/test_paddle/test_fft.py:192]: We also need to test for the default `n=None` case

**Review by hmahmood24 — CHANGES_REQUESTED** [review]

**Comment by AwkNinja** [comment]
Hi @hmahmood24 sorry for the late update but please take a look.

File: ivy_tests/test_ivy/test_frontends/test_paddle/test_fft.py [hunk 3]
```diff
@@ -187,3 +187,39 @@ def test_paddle_ifftshift(
...
+@handle_frontend_test(
+    fn_tree="paddle.fft.rfft",
+    dtype_input_axis=helpers.dtype_values_axis(
+        available_dtypes=helpers.get_dtypes("valid"),
+        min_num_dims=1,
+        min_dim_size=2,
+        shape=helpers.get_shape(min_num_dims=1,
+                                max_num_dims=2,
+                                min_dim_size=2,
+                                max_dim_size=4,),
```
Comment (hmahmood24) [inline ivy_tests/test_ivy/test_frontends/test_paddle/test_fft.py:202]: @asad0991 Perhaps you forgot to install/run `pre-commit` to fix your [lint/formatting errors](https://github.com/unifyai/ivy/actions/runs/5983491732/job/16233689761?pr=19454)?

**Review by hmahmood24 — CHANGES_REQUESTED** [review]

**Comment by AwkNinja** [comment]
@hmahmood24 Fixed the lint issues I changed devices forgot to install pre-commit on the new device.

**Review by hmahmood24 — CHANGES_REQUESTED** [review]
@AwkNinja I can see a couple of [failing tests](https://github.com/unifyai/ivy/actions/runs/6137850477/job/16654114642?pr=19454) still associated with this PR

**Comment by AwkNinja** [comment]
@hmahmood24 The tests for tensorflow gives me keyError: "cup" its like the error you get when the version is wrong in the decorator but I used the correct one. For the paddle error it just started to occur just now. Its basically in ivy.dft function I'll take a look at it.

**Comment by hmahmood24** [comment]
> @hmahmood24 The tests for tensorflow gives me keyError: "cup" its like the error you get when the version is wrong in the decorator but I used the correct one. For the paddle error it just started to occur just now. Its basically in ivy.dft function I'll take a look at it.

Pretty sure you can find a solution to that key error problem in discord since that's been brought up numerous times already

**Comment by AwkNinja** [comment]
> > @hmahmood24 The tests for tensorflow gives me keyError: "cup" its like the error you get when the version is wrong in the decorator but I used the correct one. For the paddle error it just started to occur just now. Its basically in ivy.dft function I'll take a look at it.
> 
> Pretty sure you can find a solution to that key error problem with a simple search in discord since that's been brought up numerous times already

I did read on discord chaning version to 2.5.1 and below should have fixed it. I'll have a look again might be another solution someone brought up.

**Comment by AwkNinja** [comment]
Hi @hmahmood24, the `tensorflow` tests are passing now with the change I made but It should no longer be needed once [this](https://github.com/unifyai/ivy/pull/23439) feature is implemented. Also for the paddle tests I found the issue [here](https://github.com/AwkNinja/ivy/blob/463b0af5ec4f2cf68a5d2ddb85bd5f46f5526077/ivy/functional/ivy/general.py#L2934C46-L2934C46). Basically `query_` is a `slice` object and at some point value for `query_` is `slice(0, ivy.Array(2), None)` hence giving the error `TypeError: slice indices must be integers or None or have an __index__ method`. Maybe some with more experience in the handling of `paddle` framework can look into this.

**Comment by hmahmood24** [comment]
> Hi @hmahmood24, the `tensorflow` tests are passing now with the change I made but It should no longer be needed once [this](https://github.com/unifyai/ivy/pull/23439) feature is implemented. Also for the paddle tests I found the issue [here](https://github.com/AwkNinja/ivy/blob/463b0af5ec4f2cf68a5d2ddb85bd5f46f5526077/ivy/functional/ivy/general.py#L2934C46-L2934C46). Basically `query_` is a `slice` object and at some point value for `query_` is `slice(0, ivy.Array(2), None)` hence giving the error `TypeError: slice indices must be integers or None or have an __index__ method`. Maybe someone with more experience in the handling of `paddle` framework can look into this.

Hey @AwkNinja, thanks a lot for the effort. I'll ask someone from the team to take a look at this issue you brought up with `_parse_query`. In the meanwhile I think it should be okay to merge the PR because this issue stems from `set_item` itself and not with your PR. What do you think @AnnaTz ?

**Review by hmahmood24 — APPROVED** [review]

**Comment by AnnaTz** [comment]
> > Hi @hmahmood24, the `tensorflow` tests are passing now with the change I made but It should no longer be needed once [this](https://github.com/unifyai/ivy/pull/23439) feature is implemented. Also for the paddle tests I found the issue [here](https://github.com/AwkNinja/ivy/blob/463b0af5ec4f2cf68a5d2ddb85bd5f46f5526077/ivy/functional/ivy/general.py#L2934C46-L2934C46). Basically `query_` is a `slice` object and at some point value for `query_` is `slice(0, ivy.Array(2), None)` hence giving the error `TypeError: slice indices must be integers or None or have an __index__ method`. Maybe someone with more experience in the handling of `paddle` framework can look into this.
> 
> Hey @AwkNinja, thanks a lot for the effort. I'll ask someone from the team to take a look at this issue you brought up with `_parse_query`. In the meanwhile I think it should be okay to merge the PR because this issue stems from `set_item` itself and not with your PR. What do you think @AnnaTz ?

This wasn't a `set_item` issue. I have just resolved it with a quick [fix](https://github.com/unifyai/ivy/commit/9ee0f944c2d51770f48eb0b6fec18e016d62da7e) in `ivy.dft`.

**Comment by AwkNinja** [comment]
Hi @hmahmood24 all the test cases are passing now with the fix that @AnnaTz did. Please take a look.
